back/boxtribute_server/setup_wizard.py

```diff
@@ -192,14 +199,117 @@ def _clone_products(*, source_base_id, target_base_id):
         )
 
 
+def _reconstruct(box, event):
+    if event.changes == "Box was undeleted.":
+        # The information about when the box was deleted is only available from the
+        # "Record deleted" event. Insert dummy date for now
+        box["deleted"] = "1970-01-01"
+        return True
+
+    if event.changes == "Record deleted":
+        # The box version before the deletion is obviously not deleted
+        box["deleted"] = None
+        return True
+
+    fields = {
+        "box_state_id": "state",
+        "location_id": "location",
+        "product_id": "product",
+        "size_id": "size",
+        "items": "number_of_items",
+    }
+    field = fields.get(event.changes)
+    if field is None:
+        # No relevant change happened
+        return False
+    box[field] = event.from_int
+    return True
+
+
+def _convert(*, location_id, oldest_creation_year=2020):
+    from boxtribute_server.models.definitions.box import Box
+    from boxtribute_server.models.definitions.history import DbChangeHistory
+
+    start = datetime(oldest_creation_year, 1, 1)
+    boxes = Box.select(
+        Box.id,
+        Box.state,
+        Box.created_on,
+        Box.deleted,
+        Box.number_of_items,
+        Box.location,
+        Box.product,
+        Box.size,
+    ).where(
+        Box.created_on > start,
+        Box.location == location_id,
+    )
+    LOGGER.debug(f"Selected {len(boxes)} boxes.")
+
+    # Obtain history entries for selected boxes
```

Comment:
requested a structural change to the history table, including a more complete performance trace.

Reply:
Github won't let me leave a requested changes review without a line-level comment, so this is where it is.

Reply:
@HaGuesto @pylipp

Reply:
@aerinsol current indexes of the history table are:
```sql
  KEY `user_id` (`user_id`),
  KEY `recordid_changedate_composite_index` (`record_id`,`changedate`),
  KEY `tablename_recordid_changedate_composite_index` (`tablename`,`record_id`,`changedate`),
```
This means there's already a composite index with `record_id` as leftmost index. I added an index on the `record_id` column nevertheless but there was no change in runtime.

Maybe I should stress again that the expensive part is **not** the DB select (takes about 2-3s) but the reconstruction in Python which takes 14s.

Reply:
> This means there's already a composite index with record_id as leftmost index. I added an index on the record_id column nevertheless but there was no change in runtime.

@pylipp To be clear, what I was actually proposing was **dropping all other indices except for `record_id`**, OR a composite index of `record_id `and `changedate`, as having multiple concurrent indexes creates performance issues. 

May I ask if you know where the delay in Python is coming from, if not from the processing of the SQL? If there is some other issue I don't believe that this will be solved using shadow tables as proposed in:  docs/adr/adr_backend_etl_and_hosting_for_StatViz.md

Reply:
When reconstructing, I iterate over all boxes, and then iterate over the entire history to find related entries. But obviously on average only 5 entries out of 20k are related to the box of the outer loop, resulting in a large overhead of cancelled runs of the inner loop.

This is not yet tied together with a full data computation for e.g. the `createdBoxes` query, so it's not a definite counter-argument against shadow tables.